### ISSUE TITLE
Failing to resolve p2 site entirely.

### ISSUE THREAD

**Issue by danielburrell** [reficio/p2-maven-plugin#130]
I recently followed your minimal example for bundling the commons library which worked fine. As an experiment I tried bundling `<id>org.apache.avro:avro:1.7.7</id>`.

I successfully created a site

`mvn p2:site`
`[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building example-p2-site 1.0.0
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- p2-maven-plugin:1.1.1:site (default-cli) @ example-p2-site ---
[info] Resolving artifact=[org.apache.avro:avro:1.7.7] transitive=[true] source=[false]
[info] 	 [JAR] avro-1.7.7.jar
[info] 	 [JAR] jackson-core-asl-1.9.13.jar
[info] 	 [JAR] jackson-mapper-asl-1.9.13.jar
[info] 	 [JAR] paranamer-2.3.jar
[info] 	 [JAR] snappy-java-1.0.5.jar
[info] 	 [JAR] commons-compress-1.4.1.jar
[info] 	 [JAR] xz-1.0.jar
[info] 	 [JAR] slf4j-api-1.6.4.jar
[info] Executing Bundler:
[info] 	 [SKIP] avro-1.7.7.jar
[info] Executing Bundler:
[info] 	 [SKIP] jackson-core-asl-1.9.13.jar
[info] Executing Bundler:
[info] 	 [SKIP] jackson-mapper-asl-1.9.13.jar
[info] Executing Bundler:
[info] 	 [EXEC] paranamer-2.3.jar
[warn]  : Superfluous export-package instructions: [com, com.thoughtworks]
[info] Executing Bundler:
[info] 	 [SKIP] snappy-java-1.0.5.jar
[info] Executing Bundler:
[info] 	 [SKIP] commons-compress-1.4.1.jar
[info] Executing Bundler:
[info] 	 [EXEC] xz-1.0.jar
[warn]  : Superfluous export-package instructions: [org, org.tukaani]
[info] Executing Bundler:
[info] 	 [SKIP] slf4j-api-1.6.4.jar
[info] Resolved 0features
[INFO] Command line:
`

When I went to add the p2 repo in eclipse I got

`Missing requirement: Data mapper for Jackons JSON proessor 1.9.13 (jackson-mapper-ask 1.9.13) requires 'package javax.xml.datatype 0.0.0' but it could not be found."

As I understand it, javax.xml.datatype ships with java itself. The dependency mentioned in the error message is not the avro dependency itself, it's one of the transitive dependencies  (I think avro depends on jackson mapper which in turn depends on something in that package)

What is the normal course of action to take here?

**Comment by isoubelet** [comment]
Hi Daniel,

A couple of days ago I faced the same problem with javax.jms.
The thing is that, in Eclipse, almost everything is dynamic and loaded "on the fly", even the runtime environment so all dependencies have to be defined as part of your plugin/feature information since that definition does not rely on your current underlying runtime context.
That's why plugins and features are highly portable.

Take a look at this artifact:
https://mvnrepository.com/artifact/javax.xml.parsers/jaxp-api

Try including the latest version to your p2 site...

Hope it helps!

**Comment by danielburrell** [comment]
Hmm I see, yse adding an artifact entry worked here but for example now it's asking about sun.misc 0.0.0, so then for each library in the JDK I I will end up attempting to make a bundle of it?

Is there no way to describe "trust me" or "that dependency lives in the JDK"? In the maven world this would have been declared as a 'provided' dependency in the pom file. Does OSGi not have the same concept, and if so can this plugin allow me to configure it or better yet automatically figure it out?

**Comment by isoubelet** [comment]
Yes, there is a sort a transitive dependency calculation. 
Are you writing an Eclipse plugin?

**Comment by isoubelet** [comment]
Ok, understood. So my suggestion is to slightly change the approach here, as follows:

1. Create a feature that includes all your required plugins importing them from your site (this will allow you better control in case you want to separate by group of functionalities).
2. Go to Run Configurations and create a new Eclipse Application
3. Go to plugins tab and make sure that in the combobox "Launch with" the option "all workspace and enables target plug-ing" is selected.
4. Apply and Run

This action will create an Eclipse application inheriting all the artifacts included in your host runtime environment, like org.eclipse.core.runtime and equinox.common

Could you please check that?

**Comment by danielburrell** [comment]
I don't understand, the target platform does not contain the org.apache.avro dependency because we were unable to introduce it via the p2 reference (because of the dependency failure on sun.misc). As a result if I attempt item 1 from your list, I will not see the org.apache.avro in the 'Required Plugins' section as the autocomplete list for this section is populated based on the contents of the target platform. Since the target platform has not been resolved because adding the P2 entry didn't succeed.

**Comment by isoubelet** [comment]
Yes, I got it. What I am suggesting, based on what I often do in those cases, is to add those plugins into the Eclipse you are running to create the target, then add that plugin into your target and try running an Eclipse Application.

![image](https://user-images.githubusercontent.com/25192235/37031470-2a35be24-213f-11e8-8747-bfc52e3ee0b7.png)

Then in your feature add required plugins and go to Dependencies tab and click on compute.
![image](https://user-images.githubusercontent.com/25192235/37031603-b841f6f6-213f-11e8-8557-09477e5e7515.png)

**Comment by danielburrell** [comment]
I see, so Help -> Install Software, then add my site, and add the dependencies is for screenshot 1 in order to add it to the current eclipse platform, yes?

If so, when I do this I get a similar error. "The installation cannot be completed as requested" it offers an alternative installation and a list of what can and can't be installed. In the "Will not be installed" section it shows Apache Avro. If I clock "show original error and build my own solution" I see the underlying error:

"Cannot complete the install because one or more required items could not be found.
  Software being installed: Apache Avro 1.7.7 (avro 1.7.7)
  Missing requirement: Apache Avro 1.7.7 (avro 1.7.7) requires 'package sun.misc 0.0.0' but it could not be found"

This appears to be the same error obtained when attempting to reference p2 from the target file, only this time it's from the Eclipse Install menu.

**Comment by isoubelet** [comment]
That's weird — Do you see other sites in your list like the official Eclipse repository?
If so, make sure that the option _Contact all update sites during install_ is clicked...

**Comment by danielburrell** [comment]
The Eclipse Project Updates - http://download.eclipse.org/eclipse/updates/4.7
is an entry listed. and the "Contact all update sites..." checkbox is ticked.

**Comment by isoubelet** [comment]
Try including the software repository as well: http://download.eclipse.org/releases/oxygen/

**Comment by danielburrell** [comment]
Yes Oxygen - http://download.eclipse.org/releases/oxygen was already present

**Comment by isoubelet** [comment]
Well... you could try adding this to your local site, although, at least for me, it's not the most elegant way.
<dependency>
    <groupId>com.diffplug.osgi</groupId>
    <artifactId>com.diffplug.osgi.extension.sun.misc</artifactId>
    <version>0.0.0</version>
</dependency>

**Comment by escosoft** [comment]
Hi, I'm greatly interested in this issue as we are facing the exact same issue and we are unable to solve it

**Comment by danielburrell** [comment]
ok so I tried that and got a little further, but I'm now quite interested in a checkbox i found which says "Include required software" "By default all required software is added to the target based on its environment settings. Turning this option off allows software to be added with missing requirements and multiple environment. This setting applies to the entire target definition."

So I'm thinking as long as I know the dependencies that I *actually* need, I might be able to get away with this option. Going to try it with just avro and see if the project is tricked into running.

Anybody advise on the risks associated with this? Am I offsetting compile time errors to potential runtime errors? I say potential because presumably if the missing class isn't loaded because the feature that would cause it to be invoked isn't used then I'm ok right>?
